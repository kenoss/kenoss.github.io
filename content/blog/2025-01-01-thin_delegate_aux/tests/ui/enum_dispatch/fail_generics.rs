struct Rect<T> {
    width: T,
    height: T,
}

#[enum_dispatch::enum_dispatch]
enum Shape<T> {
    Rect(Rect<T>),
}

#[enum_dispatch::enum_dispatch(Shape<usize>)]
trait ShapeI {
    fn area(&self) -> usize;
}

impl<T> ShapeI for Rect<T>
where
    T: std::ops::Mul,
{
    fn area(&self) -> usize {
        (self.width * self.height) as usize
    }
}

fn main() {
    let rect = Rect { width: 2, height: 3 };
    assert_eq!(rect.area(), 6);
    let shape = Shape::Rect(rect);
    assert_eq!(shape.area(), 6);
}

// The result of `cargo expand`:
//
// #![feature(prelude_import)]
// #[prelude_import]
// use std::prelude::rust_2021::*;
// #[macro_use]
// extern crate std;
// struct Rect<T> {
//     width: T,
//     height: T,
// }
// enum Shape<T> {
//     Rect(Rect<T>),
// }
// trait ShapeI {
//     fn area(&self) -> usize;
// }
// impl<T> ::core::convert::From<Rect<T>> for Shape<T> {
//     fn from(v: Rect<T>) -> Shape<T> {
//         Shape::Rect(v)
//     }
// }
// impl<T> ::core::convert::TryInto<Rect<T>> for Shape<T> {
//     type Error = &'static str;
//     fn try_into(
//         self,
//     ) -> ::core::result::Result<
//         Rect<T>,
//         <Self as ::core::convert::TryInto<Rect<T>>>::Error,
//     > {
//         match self {
//             Shape::Rect(v) => Ok(v),
//         }
//     }
// }
// impl<T> ShapeI for Shape<T> {
//     #[inline]
//     fn area(&self) -> usize {
//         match self {
//             Shape::Rect(inner) => ShapeI::area(inner),
//         }
//     }
// }
// impl<T> ShapeI for Rect<T>
// where
//     T: std::ops::Mul,
// {
//     fn area(&self) -> usize {
//         (self.width * self.height) as usize
//     }
// }
// fn main() {
//     let rect = Rect { width: 2, height: 3 };
//     match (&rect.area(), &6) {
//         (left_val, right_val) => {
//             if !(*left_val == *right_val) {
//                 let kind = ::core::panicking::AssertKind::Eq;
//                 ::core::panicking::assert_failed(
//                     kind,
//                     &*left_val,
//                     &*right_val,
//                     ::core::option::Option::None,
//                 );
//             }
//         }
//     };
//     let shape = Shape::Rect(rect);
//     match (&shape.area(), &6) {
//         (left_val, right_val) => {
//             if !(*left_val == *right_val) {
//                 let kind = ::core::panicking::AssertKind::Eq;
//                 ::core::panicking::assert_failed(
//                     kind,
//                     &*left_val,
//                     &*right_val,
//                     ::core::option::Option::None,
//                 );
//             }
//         }
//     };
// }
