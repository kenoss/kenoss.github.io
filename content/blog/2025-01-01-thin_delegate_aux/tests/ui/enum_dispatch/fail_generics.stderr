error[E0277]: cannot multiply `T` by `T`
  --> tests/ui/enum_dispatch/fail_generics.rs:13:5
   |
12 |   trait ShapeI {
   |  _______-
13 | |     fn area(&self) -> usize;
   | |_____^^____- required by a bound introduced by this call
   |       |
   |       no implementation for `T * T`
   |
note: required for `Rect<T>` to implement `ShapeI`
  --> tests/ui/enum_dispatch/fail_generics.rs:16:9
   |
16 | impl<T> ShapeI for Rect<T>
   |         ^^^^^^     ^^^^^^^
17 | where
18 |     T: std::ops::Mul,
   |        ------------- unsatisfied trait bound introduced here
help: consider restricting type parameter `T`
   |
11 | #[enum_dispatch::enum_dispatch(Shape<usize>)]: std::ops::Mul
   |                                              +++++++++++++++

error[E0605]: non-primitive cast: `<T as Mul>::Output` as `usize`
  --> tests/ui/enum_dispatch/fail_generics.rs:21:9
   |
21 |         (self.width * self.height) as usize
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
